ApplicationLibCode/ProjectDataModel/EnsembleFileSet/RimEnsembleFileSet.cpp

```diff
@@ -0,0 +1,366 @@
+    }
+    else
+    {
+        templateText = m_nameTemplateString();
+    }
+
+    std::map<QString, QString> keyValues = {
+        { RiaDefines::key1VariableName(), QString::fromStdString( key1 ) },
+        { RiaDefines::key2VariableName(), QString::fromStdString( key2 ) },
+    };
+
+    auto candidateName = RiaTextStringTools::replaceTemplateTextWithValues( templateText, keyValues );
+
+    if ( m_autoName )
+    {
+        candidateName = candidateName.trimmed();
+
+        // When using auto name, remove leading and trailing commas that may occur if key1 or key2 is empty
+        if ( candidateName.startsWith( "," ) )
+        {
+            candidateName = candidateName.mid( 1 );
+        }
+        if ( candidateName.endsWith( "," ) )
+        {
+            candidateName = candidateName.left( candidateName.length() - 1 );
+        }
+
+        auto generateUniqueName = []( const QString& baseName, const std::set<QString>& existingNames )
+        {
+            // Avoid identical ensemble names by appending a number
+            QString uniqueName = baseName;
+            int     counter    = 1;
+            while ( existingNames.find( uniqueName ) != existingNames.end() )
+            {
+                uniqueName = QString( "%1_%2" ).arg( baseName ).arg( counter++ );
+            }
+            return uniqueName;
+        };
+
+        candidateName = generateUniqueName( candidateName, existingEnsembleNames );
+    }
+
+    if ( name() == candidateName ) return;
+
+    setName( candidateName );
+    nameChanged.send();
+}
+
+//--------------------------------------------------------------------------------------------------
+///
+//--------------------------------------------------------------------------------------------------
+void RimEnsembleFileSet::setUsePathKey1( bool useKey1 )
+{
+    m_useKey1 = useKey1;
+}
+
+//--------------------------------------------------------------------------------------------------
+///
+//--------------------------------------------------------------------------------------------------
+void RimEnsembleFileSet::setUsePathKey2( bool useKey2 )
+{
+    m_useKey2 = useKey2;
+}
+
+//--------------------------------------------------------------------------------------------------
+///
+//--------------------------------------------------------------------------------------------------
+void RimEnsembleFileSet::defineEditorAttribute( const caf::PdmFieldHandle* field, QString uiConfigName, caf::PdmUiEditorAttribute* attribute )
+{
+    if ( field == &m_pathPattern )
+    {
+        if ( auto lineEdAttr = dynamic_cast<caf::PdmUiLineEditorAttribute*>( attribute ) )
+        {
+            lineEdAttr->placeholderText = "Enter path pattern...";
+        }
+    }
+    else if ( field == &m_realizationSubSet )
+    {
+        if ( auto lineEdAttr = dynamic_cast<caf::PdmUiLineEditorAttribute*>( attribute ) )
+        {
+            lineEdAttr->placeholderText = "E.g. 0,1,2-5,10-20";
```

Comment:
Maybe 4-6 instead of 2-5 in example pattern, since it could be more effectively written as "0-5, 10-20".